core/data.go

```diff
@@ -483,34 +486,80 @@ func (cb Bundles) FromEncodedBundles(eb EncodedBundles) (Bundles, error) {
 
 // PaymentMetadata represents the header information for a blob
 type PaymentMetadata struct {
-	// Existing fields
-	AccountID string
+	// AccountID is the ETH account address for the payer
+	AccountID string `json:"account_id"`
 
-	// New fields
-	BinIndex uint32
+	// BinIndex represents the range of time at which the dispersal is made
+	BinIndex uint32 `json:"bin_index"`
 	// TODO: we are thinking the contract can use uint128 for cumulative payment,
 	// but the definition on v2 uses uint64. Double check with team.
-	CumulativePayment *big.Int
+	CumulativePayment *big.Int `json:"cumulative_payment"`
 }
 
 // Hash returns the Keccak256 hash of the PaymentMetadata
-func (pm *PaymentMetadata) Hash() []byte {
-	// Create a byte slice to hold the serialized data
-	data := make([]byte, 0, len(pm.AccountID)+4+pm.CumulativePayment.BitLen()/8+1)
-
-	// Append AccountID
-	data = append(data, []byte(pm.AccountID)...)
-
-	// Append BinIndex
-	binIndexBytes := make([]byte, 4)
-	binary.BigEndian.PutUint32(binIndexBytes, pm.BinIndex)
-	data = append(data, binIndexBytes...)
-
-	// Append CumulativePayment
-	paymentBytes := pm.CumulativePayment.Bytes()
-	data = append(data, paymentBytes...)
+func (pm *PaymentMetadata) Hash() ([32]byte, error) {
+	blobHeaderType, err := abi.NewType("tuple", "", []abi.ArgumentMarshaling{
+		{
+			Name: "accountID",
+			Type: "string",
+		},
+		{
+			Name: "binIndex",
+			Type: "uint32",
+		},
+		{
+			Name: "cumulativePayment",
+			Type: "uint256",
+		},
+	})
+	if err != nil {
+		return [32]byte{}, err
+	}
+
+	arguments := abi.Arguments{
+		{
+			Type: blobHeaderType,
+		},
+	}
+
+	bytes, err := arguments.Pack(pm)
+	if err != nil {
+		return [32]byte{}, err
+	}
+
+	var hash [32]byte
+	hasher := sha3.NewLegacyKeccak256()
+	hasher.Write(bytes)
+	copy(hash[:], hasher.Sum(nil)[:32])
+
+	return hash, nil
+}
+
+func (pm *PaymentMetadata) MarshalDynamoDBAttributeValue() (types.AttributeValue, error) {
```

Comment:
Does this belong in core? Is there any way to have this code in the blobstore?

Reply:
They need to belong in the same package as where the type is defined. We can wrap and clone these types in blobstore but not sure if that's worth it